blog/2024-01-24-transforming-your-data-1/index.md

```diff
@@ -0,0 +1,472 @@
+as the edge table requires. To do this, `transaction.csv` file isn't loaded as-is, but is instead
+transformed into the `transacted_with.csv` file via the following query.
+
+```sql
+-- Generate `transacted_with.csv` from `transaction.csv`
+COPY (
+    LOAD FROM 'transaction.csv' (header=true)
+    RETURN
+        client_id,
+        merchant_id,
+        amount_usd,
+        timestamp
+)
+TO 'transacted_with.csv';
+```
+
+The example above consists of two subquery blocks: `LOAD` and `COPY`. The `LOAD` block is used to
+scan the CSV file, check for headers and data types and return the columns we need. The `COPY`
+block is used to write the results we obtained from the scan to a new file.
+
+We can do the same for the other two edge tables as well.
+
+```sql
+// Generate `belongs_to.csv` from `merchant.csv`
+COPY (
+    LOAD FROM 'merchant.csv' (header=true)
+    RETURN
+        merchant_id,
+        company_id
+)
+TO 'belongs_to.csv';
+```
+
+```sql
+// Generate `located_in.csv` from `merchant.csv`
+COPY (
+    LOAD FROM 'merchant.csv' (header=true)
+    RETURN
+        merchant_id,
+        city_id
+)
+TO 'located_in.csv';
+```
+
+With all the input files in place, we can now proceed to insert the data and build the graph!
+
+## Insert data into Kùzu
+
+Collecting all the above functions, we can write a script that performs the following:
+
+1. Transform node/edge tables as needed
+1. Create the node tables
+1. Create edge tables
+1. Load the node tables into the database
+1. Load the edge tables into the database
+
+In Python, it would look something like this:
+
+```sql
+// Copy from CSV to node tables
+COPY Client FROM 'client.csv' (header=true)
+COPY City FROM 'city.csv' (header=true)
+COPY Company FROM 'company.csv' (header=true)
+COPY Merchant FROM 'merchant.csv' (header=true)
+
+// Copy from CSV to edge tables
+COPY TransactedWith FROM 'transacted_with.csv' 
+COPY BelongsTo FROM 'belongs_to.csv'
+COPY LocatedIn FROM 'located_in.csv'
+```
+
+The queries above require that the empty tables were created beforehand. The `COPY <edge_table> FROM <file>` statement
+writes the data into a Kùzu database. Running the queries on an existing database connection
+results in the graph being saved to a local directory.
+
+## Querying the graph
+
+We then run some simple queries to test that the data was loaded correctly. We can either create
+a standalone script using the client SDK of your choice, or fire up a [Kùzu CLI](https://kuzudb.com/docusaurus/getting-started/cli)
+shell and run some Cypher queries.
+
+The first query showcases specifically how Kùzu is beneficial in expressing query logic that
+requires traversing multiple edges. In this query, we find all clients that have transacted with
+at least one of the merchants with IDs 2 and 11.
+
+```sql
+// Who are the clients that made transactions in at least one of the merchants with IDs 2 and 11?
```

Comment:
As per our discussion, this query looks off. We can do the 3 merchant case. Or one of the others.